src/esp/gfx/PBRShader.h

```diff
@@ -0,0 +1,348 @@
+   *  light;
+   */
+  PBRShader& setLightPosition(unsigned int lightIndex,
+                              const Magnum::Vector3& pos);
+
+  /**
+   *  @brief Set the direction of a specific light.
+   *  @param lightIndex, the index of the light, MUST be smaller than
+   * lightCount_
+   *  @param dir, the direction of the light in *camera* space
+   *  @return Reference to self (for method chaining)
+   *  NOTE:
+   *  If the light was a point light, it will be overrided as a direction
+   * light;
+   */
+  PBRShader& setLightDirection(unsigned int lightIndex,
+                               const Magnum::Vector3& dir);
+
+  /**
+   *  @brief Set the range of a specific light.
+   *  @param lightIndex, the index of the light, MUST be smaller than
+   * lightCount_
+   *  @param range, the range of the light
+   *  @return Reference to self (for method chaining)
+   */
+  PBRShader& setLightRange(unsigned int lightIndex, float range);
+
+  /**
+   *  @brief Set the color of a specific light.
+   *  @param lightIndex, the index of the light, MUST be smaller than
+   * lightCount_
+   *  @param color, the color of the light
+   *  @return Reference to self (for method chaining)
+   */
+  PBRShader& setLightColor(unsigned int lightIndex,
+                           const Magnum::Vector3& color,
+                           float intensity = 1.0);
+
+ protected:
+  Flags flags_;
+  unsigned int lightCount_;
+
+  // ======= uniforms =======
+  // it hurts the performance to call glGetUniformLocation() every frame due
+  // to string operations. therefore, cache the locations in the constructor
+  // material uniforms
+  int modelviewMatrixUniform_ = ID_UNDEFINED;
+  int normalMatrixUniform_ = ID_UNDEFINED;
+  int mvpMatrixUniform_ = ID_UNDEFINED;
+  int baseColorUniform_ = ID_UNDEFINED;  // diffuse color
+  int roughnessUniform_ = ID_UNDEFINED;  // roughness of a surface
+  int metallicUniform_ = ID_UNDEFINED;
+  int albedoTextureUniform_ = ID_UNDEFINED;
+  int roughnessTextureUniform_ = ID_UNDEFINED;
+  int metallicTextureUniform_ = ID_UNDEFINED;
+  int normalTextureUniform_ = ID_UNDEFINED;
+  int objectIdUniform_ = ID_UNDEFINED;
+
+  struct LightUniformLocations {
+    int color = ID_UNDEFINED;
+    int range = ID_UNDEFINED;
+  };
+  std::vector<LightUniformLocations> lightsUniform_;
+
+  // Light direction is not in the LightUniformLocations
+  // simply because the light direction might be used in both .vert and .frag
+  // but the other info is only used in .frag;
+  // And MOST importantly, the lightDirection (in camera space) depends on
+  // camera matrix so it changes all the time, but the other info does not.
+  // NOTE:
+  // In fragement shader, the "LightDirection" is a vec4.
+  // when w == 0, it means .xyz is the light direction;
+  // when w == 1, it means it is the light position, NOT the direction;
+  std::vector<int> lightDirectionsUniform_;
+};
+
+CORRADE_ENUMSET_OPERATORS(PBRShader::Flags);
```

Comment:
```suggestion
CORRADE_ENUMSET_OPERATORS(PBRShader::Flags)
```

Reply:
By the way, what kind of out-of-class operators would I have after I defined this?

Reply:
The [bitwise operators from here](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1EnumSet.html#pub-methods).

Hmm, now you got me thinking, I think there's even a way to not need this macros at all. (This is originally modelled after Qt, which also had such a macro, so I didn't question this at all... but now I think in-class-defined friends could work as well?)